src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2281,46 +2279,47 @@ void ApplyPreserveInfo (TypeDefinition type)
 		{
 			ApplyPreserveMethods (type);
 
-			if (!Annotations.TryGetPreserve (type, out TypePreserve preserve))
-				return;
-
-			var di = new DependencyInfo (DependencyKind.TypePreserve, type);
+			if (Annotations.TryGetPreserve (type, out TypePreserve preserve)) {
+				var di = new DependencyInfo (DependencyKind.TypePreserve, type);
 
-			switch (preserve) {
-			case TypePreserve.All:
-				Annotations.TryGetPreserveAccessibility (type, out preserve);
 				switch (preserve) {
-				case TypePreserve.AccessibilityVisible:
+				case TypePreserve.All:
+					MarkFields (type, true, di);
+					MarkMethods (type, di, type);
+					return;
+
+				case TypePreserve.Fields:
+					if (!MarkFields (type, true, di, true))
+						_context.LogWarning ($"Type {type.GetDisplayName ()} has no fields to preserve", 2001, type);
+					break;
+				case TypePreserve.Methods:
+					if (!MarkMethods (type, di, type))
+						_context.LogWarning ($"Type {type.GetDisplayName ()} has no methods to preserve", 2002, type);
+					break;
+				}
+			}
+
+			if (Annotations.TryGetPreservedMembers (type, out TypePreserveMembers members)) {
+				var di = new DependencyInfo (DependencyKind.TypePreserve, type);
```

Comment:
I guess I don't understand how the type preserve and type preserve members should interact. It seems that:
* `All` -> all regardless of members visibility (and regardless of `AllVisible` or `AllVisibleOrInternal`).
* `Methods` -> all methods regardless of visibility - but then we also apply `AllVisible` or `AllVisibleOrInternal` on top of that.
* `Fields` -> similar to `Methods` but with fields.
* `None` -> still apply `AllVisible` or `AllVisibleOrInternal`.

Is that the desired behavior?

Reply:
So to simplify the two settings are orthogonal - we apply the union of both. I think it would make sense to somehow change the naming - they seem related since the names are so similar. Or maybe it would make sense to unify them into one thing - something like:
* `AllMethods`, `VisibleMethods`, `VisibleOrInternalMethods`
* `AllFields`, `VisibleFields`, `VisibleOrInternalFields`

And then `preserve=all` would be set as `AllMethods | AllFields`, `preserve=methods` would mean `AllMethods` and marking an assembly as `visible` would add `VisibleMethods | VisibleFields` to all types in the assembly.

Then the implementation here would be straightforward...

Reply:
Or at least add a comment about the intended behavior somewhere :-)

Reply:
> Then the implementation here would be straightforward

I tried to unify them because I thought it'd be straight forward but had to undo that because it's too complex to do the merge and not worth the code complexity. For example when end up preserving All Methods and Visible Fields. I'd have to change existing enum to have only flags that could break custom steps.

Reply:
Right - I didn't realize that the enum is publicly visible...
OK - let's keep it separate.